File: cathode/src/main/java/net/simonvt/cathode/ui/movie/MovieHistoryFragment.java
/*
 * Copyright (C) 2017 Simon Vig Therkildsen
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package net.simonvt.cathode.ui.movie;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.LinearLayout;
import android.widget.TextView;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.lifecycle.Observer;
import androidx.lifecycle.ViewModelProviders;
import butterknife.BindView;
import dagger.android.support.AndroidSupportInjection;
import java.util.ArrayList;
import java.util.List;
import java.util.Locale;
import javax.inject.Inject;
import net.simonvt.cathode.R;
import net.simonvt.cathode.api.service.SyncService;
import net.simonvt.cathode.common.ui.fragment.RefreshableAppBarFragment;
import net.simonvt.cathode.common.util.Ids;
import net.simonvt.cathode.common.util.guava.Preconditions;
import net.simonvt.cathode.entity.Movie;
import net.simonvt.cathode.images.ImageType;
import net.simonvt.cathode.images.ImageUri;
import net.simonvt.cathode.provider.helper.MovieDatabaseHelper;
import net.simonvt.cathode.sync.scheduler.MovieTaskScheduler;

public class MovieHistoryFragment extends RefreshableAppBarFragment {

  private static final String TAG = "net.simonvt.cathode.ui.show.MovieHistoryFragment";

  private static final String ARG_MOVIEID = TAG + ".movieId";
  private static final String ARG_MOVIETITLE = TAG + ".movieTitle";

  enum Type {
    LOADING, ERROR, EMPTY, ITEM
  }

  private Movie movie;

  private MovieHistoryLiveData.Result result;

  @Inject MovieTaskScheduler movieScheduler;
  @Inject SyncService syncService;
  @Inject MovieDatabaseHelper movieHelper;

  private long movieId;
  private String movieTitle;

  @Inject MovieHistoryViewModelFactory viewModelFactory;
  private MovieHistoryViewModel viewModel;

  @BindView(R.id.topTitle) TextView released;
  @BindView(R.id.topSubtitle) TextView rating;

  @BindView(R.id.watchedAtContainer) LinearLayout watchedAtContainer;

  public static String getTag(long movieId) {
    return TAG + "/" + movieId + "/history/" + Ids.newId();
  }

  public static Bundle getArgs(long movieId, String movieTitle) {
    Preconditions.checkArgument(movieId >= 0, "movieId must be >= 0, was " + movieId);

    Bundle args = new Bundle();
    args.putLong(ARG_MOVIEID, movieId);
    args.putString(ARG_MOVIETITLE, movieTitle);
    return args;
  }

  @Override public void onCreate(@Nullable Bundle inState) {
    super.onCreate(inState);
    AndroidSupportInjection.inject(this);

    Bundle args = getArguments();
    movieId = args.getLong(ARG_MOVIEID);
    movieTitle = args.getString(ARG_MOVIETITLE);
    setTitle(movieTitle);

    viewModel = ViewModelProviders.of(this, viewModelFactory).get(MovieHistoryViewModel.class);
    viewModel.setMovieId(movieId);
    viewModel.getMovie().observe(this, new Observer<Movie>() {
      @Override public void onChanged(Movie movie) {
        setMovie(movie);
      }
    });
    viewModel.getHistory().observe(this, new Observer<MovieHistoryLiveData.Result>() {
      @Override public void onChanged(MovieHistoryLiveData.Result result) {
        setResult(result);
        setRefreshing(false);
      }
    });
  }

  @Override
  protected View createView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container,
      @Nullable Bundle inState) {
    return inflater.inflate(R.layout.history_fragment, container, false);
  }

  @Override public void onViewCreated(@NonNull View view, @Nullable Bundle inState) {
    super.onViewCreated(view, inState);
    setResult(result);
  }

  public void onRemoveHistoryItem(long historyId, int position) {
    result.getItems().remove(position);
    setResult(result);

    movieScheduler.removeHistoryItem(movieId, historyId, result.getItems().size() == 0);
  }

  @Override public void onRefresh() {
    viewModel.getHistory().loadData();
  }

  private void setMovie(Movie movie) {
    this.movie = movie;
    movieTitle = movie.getTitle();

    final String backdropUri = ImageUri.create(ImageUri.ITEM_MOVIE, ImageType.BACKDROP, movieId);
    setTitle(movieTitle);
    setBackdrop(backdropUri);
    this.released.setText(movie.getReleased());
    this.rating.setText(String.format(Locale.getDefault(), "%.1f", movie.getRating()));
  }

  private boolean typeOrClear(ViewGroup parent, Type type) {
    if (parent.getChildCount() > 0) {
      View firstChild = parent.getChildAt(0);
      final Type childType = (Type) firstChild.getTag();
      if (childType == type) {
        return true;
      } else {
        parent.removeAllViews();
      }
    }

    return false;
  }

  private void setResult(MovieHistoryLiveData.Result result) {
    this.result = result;

    if (getView() == null) {
      return;
    }

    if (result == null) {
      if (typeOrClear(watchedAtContainer, Type.LOADING)) {
        return;
      }

      View v = LayoutInflater.from(watchedAtContainer.getContext())
          .inflate(R.layout.history_progress, watchedAtContainer, false);
      v.setTag(Type.LOADING);
      watchedAtContainer.addView(v);
    } else if (!result.isSuccessful()) {
      if (typeOrClear(watchedAtContainer, Type.ERROR)) {
        return;
      }

      View v = LayoutInflater.from(watchedAtContainer.getContext())
          .inflate(R.layout.history_error, watchedAtContainer, false);
      v.setTag(Type.ERROR);
      watchedAtContainer.addView(v);
    } else if (result.getItems().size() == 0) {

      if (typeOrClear(watchedAtContainer, Type.EMPTY)) {
        return;
      }

      View v = LayoutInflater.from(watchedAtContainer.getContext())
          .inflate(R.layout.history_empty, watchedAtContainer, false);
      v.setTag(Type.ERROR);
      watchedAtContainer.addView(v);
    } else {
      List<HistoryItem> items = result.getItems();
      List<Long> ids = new ArrayList<>();
      for (HistoryItem item : items) {
        ids.add(item.historyId);
      }

      for (int i = watchedAtContainer.getChildCount() - 1; i >= 0; i--) {
        View v = watchedAtContainer.getChildAt(i);
        if (v.getTag() != Type.ITEM) {
          watchedAtContainer.removeViewAt(i);
          continue;
        }

        final long id = (long) v.getTag(R.id.historyId);
        if (!ids.contains(id)) {
          watchedAtContainer.removeViewAt(i);
        }
      }

      for (int i = 0, size = items.size(); i < size; i++) {
        final HistoryItem item = items.get(i);
        final int position = i;

        View v = null;

        if (i < watchedAtContainer.getChildCount()) {
          v = watchedAtContainer.getChildAt(i);
        }

        if (v != null) {
          final long id = (long) v.getTag(R.id.historyId);

          if (item.historyId == id) {
            TextView watchedAt = v.findViewById(R.id.watchedAt);
            View remove = v.findViewById(R.id.remove);

            watchedAt.setText(item.watchedAt);
            remove.setOnClickListener(new View.OnClickListener() {
              @Override public void onClick(View v) {
                onRemoveHistoryItem(item.historyId, position);
              }
            });

            continue;
          }
        }

        v = LayoutInflater.from(watchedAtContainer.getContext())
            .inflate(R.layout.history_row, watchedAtContainer, false);
        TextView watchedAt = v.findViewById(R.id.watchedAt);
        View remove = v.findViewById(R.id.remove);

        v.setTag(Type.ITEM);
        v.setTag(R.id.historyId, item.historyId);
        watchedAt.setText(item.watchedAt);
        remove.setOnClickListener(new View.OnClickListener() {
          @Override public void onClick(View v) {
            onRemoveHistoryItem(item.historyId, position);
          }
        });

        watchedAtContainer.addView(v, i);
      }
    }
  }
}

File: cathode/src/main/java/net/simonvt/cathode/ui/movie/MovieHistoryFragment.kt
/*
 * Copyright (C) 2017 Simon Vig Therkildsen
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package net.simonvt.cathode.ui.movie

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.viewModels
import androidx.lifecycle.Observer
import net.simonvt.cathode.R
import net.simonvt.cathode.common.ui.fragment.RefreshableAppBarFragment
import net.simonvt.cathode.common.util.Ids
import net.simonvt.cathode.common.util.guava.Preconditions
import net.simonvt.cathode.databinding.HistoryFragmentBinding
import net.simonvt.cathode.databinding.HistoryRowBinding
import net.simonvt.cathode.entity.Movie
import net.simonvt.cathode.images.ImageType.BACKDROP
import net.simonvt.cathode.images.ImageUri
import net.simonvt.cathode.sync.scheduler.MovieTaskScheduler
import net.simonvt.cathode.ui.CathodeViewModelFactory
import net.simonvt.cathode.ui.movie.MovieHistoryFragment.Type.EMPTY
import net.simonvt.cathode.ui.movie.MovieHistoryFragment.Type.ERROR
import net.simonvt.cathode.ui.movie.MovieHistoryFragment.Type.ITEM
import net.simonvt.cathode.ui.movie.MovieHistoryFragment.Type.LOADING
import net.simonvt.cathode.ui.movie.MovieHistoryLiveData.Result
import java.util.ArrayList
import java.util.Locale
import javax.inject.Inject

class MovieHistoryFragment @Inject constructor(
  private val movieScheduler: MovieTaskScheduler,
  private val viewModelFactory: CathodeViewModelFactory
) : RefreshableAppBarFragment() {

  internal enum class Type {
    LOADING, ERROR, EMPTY, ITEM
  }

  private var movie: Movie? = null
  private var result: Result? = null
  private var movieId: Long = 0
  private var movieTitle: String? = null
  private val viewModel: MovieHistoryViewModel by viewModels { viewModelFactory }

  private var _binding: HistoryFragmentBinding? = null
  private val binding get() = _binding!!

  override fun onCreate(inState: Bundle?) {
    super.onCreate(inState)
    movieId = requireArguments().getLong(ARG_MOVIEID)
    movieTitle = requireArguments().getString(ARG_MOVIETITLE)
    setTitle(movieTitle)

    viewModel.setMovieId(movieId)
    viewModel.movie!!.observe(this, Observer { movie -> setMovie(movie) })
    viewModel.history!!.observe(
      this,
      Observer { result ->
        setResult(result)
        setRefreshing(false)
      })
  }

  override fun createView(
    inflater: LayoutInflater,
    container: ViewGroup?,
    inState: Bundle?
  ): View? {
    _binding = HistoryFragmentBinding.inflate(inflater, container, false)
    return binding.root
  }

  override fun onViewCreated(view: View, inState: Bundle?) {
    super.onViewCreated(view, inState)
    setResult(result)
  }

  override fun onDestroyView() {
    _binding = null
    super.onDestroyView()
  }

  fun onRemoveHistoryItem(historyId: Long, position: Int) {
    result!!.getItems().removeAt(position)
    setResult(result)
    movieScheduler.removeHistoryItem(movieId, historyId, result!!.getItems().size == 0)
  }

  override fun onRefresh() {
    viewModel.history!!.loadData()
  }

  private fun setMovie(movie: Movie) {
    this.movie = movie
    movieTitle = movie.title
    val backdropUri = ImageUri.create(ImageUri.ITEM_MOVIE, BACKDROP, movieId)
    setTitle(movieTitle)
    setBackdrop(backdropUri)
    binding.top.topTitle.text = movie.released
    binding.top.topSubtitle.text = String.format(Locale.getDefault(), "%.1f", movie.rating)
  }

  private fun typeOrClear(parent: ViewGroup, type: Type): Boolean {
    if (parent.childCount > 0) {
      val firstChild = parent.getChildAt(0)
      val childType = firstChild.getTag(R.id.historyType) as Type
      if (childType == type) {
        return true
      } else {
        parent.removeAllViews()
      }
    }
    return false
  }

  private fun setResult(result: Result?) {
    this.result = result
    if (view == null) {
      return
    }
    if (result == null) {
      if (typeOrClear(binding.content.content, LOADING)) {
        return
      }
      val v = LayoutInflater.from(binding.content.content.context)
        .inflate(R.layout.history_progress, binding.content.content, false)
      v.setTag(R.id.historyType, LOADING)
      binding.content.content.addView(v)
    } else if (!result.isSuccessful()) {
      if (typeOrClear(binding.content.content, ERROR)) {
        return
      }
      val v = LayoutInflater.from(binding.content.content.context)
        .inflate(R.layout.history_error, binding.content.content, false)
      v.setTag(R.id.historyType, ERROR)
      binding.content.content.addView(v)
    } else if (result.getItems().size == 0) {
      if (typeOrClear(binding.content.content, EMPTY)) {
        return
      }
      val v = LayoutInflater.from(binding.content.content.context)
        .inflate(R.layout.history_empty, binding.content.content, false)
      v.setTag(R.id.historyType, ERROR)
      binding.content.content.addView(v)
    } else {
      val items = result.getItems()
      val ids: MutableList<Long> = ArrayList()
      for (item in items) {
        ids.add(item.historyId)
      }
      for (i in binding.content.content.childCount - 1 downTo 0) {
        val v = binding.content.content.getChildAt(i)
        if (v.getTag(R.id.historyType) !== ITEM) {
          binding.content.content.removeViewAt(i)
          continue
        }
        val id = v.getTag(R.id.historyId) as Long
        if (!ids.contains(id)) {
          binding.content.content.removeViewAt(i)
        }
      }
      var i = 0
      val size = items.size
      while (i < size) {
        val item = items[i]
        val position = i
        var v: View? = null
        if (i < binding.content.content.childCount) {
          v = binding.content.content.getChildAt(i)
        }
        if (v != null) {
          val id = v.getTag(R.id.historyId) as Long
          if (item.historyId == id) {
            val itemBinding = v.tag as HistoryRowBinding
            itemBinding.watchedAt.text = item.watchedAt
            itemBinding.remove.setOnClickListener { onRemoveHistoryItem(item.historyId, position) }
            i++
            continue
          }
        }
        val itemBinding = HistoryRowBinding.inflate(
          LayoutInflater.from(binding.content.content.context),
          binding.content.content,
          false
        )
        itemBinding.root.tag = itemBinding
        itemBinding.root.setTag(R.id.historyType, ITEM)
        itemBinding.root.setTag(R.id.historyId, item.historyId)
        itemBinding.watchedAt.text = item.watchedAt
        itemBinding.remove.setOnClickListener { onRemoveHistoryItem(item.historyId, position) }
        binding.content.content.addView(itemBinding.root, i)
        i++
      }
    }
  }

  companion object {
    private const val TAG = "net.simonvt.cathode.ui.show.MovieHistoryFragment"
    private const val ARG_MOVIEID = "$TAG.movieId"
    private const val ARG_MOVIETITLE = "$TAG.movieTitle"
    fun getTag(movieId: Long): String {
      return TAG + "/" + movieId + "/history/" + Ids.newId()
    }

    fun getArgs(movieId: Long, movieTitle: String?): Bundle {
      Preconditions.checkArgument(
        movieId >= 0,
        "movieId must be >= 0, was $movieId"
      )
      val args = Bundle()
      args.putLong(ARG_MOVIEID, movieId)
      args.putString(ARG_MOVIETITLE, movieTitle)
      return args
    }
  }
}
